open-company/open-company-auth
# Invite throttle

#### Card: https://trello.com/c/YBJz95A6

#### Related PRs (review together):
- [Web PR](https://github.com/open-company/open-company-web/pull/1182)
- [Lib PR](https://github.com/open-company/open-company-lib/pull/88)
- [Support PR](https://github.com/open-company/open-company-support/pull/1)

#### Description of the change:
To prevent future suspension of our SES account we want to improve even more the limitation around user invitation.
These PRs are meant to add a limit of the number of invites a user can send in a certain amount of time and also add a CSRF token that will prevent cross-site calls of our invite API.

The whole thing is based on a Dynamodb table that has user-id as key, team-id as range key, then it contains a token column that is the CSRF token and number of invites sent in the current period.

The ttl will make sure this record expire when we want so the count will be reset to 0. The keys will make sure the limitation is applied by user/team not to a certain user on all the teams he has (no need to test or get into this, we don't have multiple teams customers right now).

The CSRF token is refreshed every time the user request the invite link, that means only 1 session at the time will be able to send invites.

NB: all this is on staging but with lower limits for easy tests. On staging you can send 10 invites every 10 minutes. On production it's 100 invites every hour. You can test this on staging if you want, just remember to use real emails (or at least emails that seems real because there is also the recipient validation PR deployed there which will fail inviting users with fake emails).

To test:
- sign in with your stoatlabs Slack account
- [ ] try inviting a user: did it work? Good
- now invite another 9 users in less than 10 minutes
- invite another user
- [ ] did it fail? Good
- wait 10 minutes from the first invite sent
- try inviting another user now
- now fill in the email of a user to invite
- copy the url of the current page
- open a new tab and paste it (use the same browser session)
- open the invite panel
- go back to the initial tab and try inviting
- [ ] did it fail? Good (if it doesn't fails it means the client refreshed the team data which refresh the CSRF token, sometimes that happens automatically depending on a few factors)

This should be it, please drop me a note when this is merged since I need to do a bit of cleanup in the infra repo (deployed branch mostly).
